### PR TITLE
Auto series creation

### PR DESCRIPTION
**Pull request by siren-tamer**
Allows a buyer to create a new series if it doesn't already exist.

* Buyer specifies the existing deployed AMM to use (which contains underlying, price, and collateral token)
* New strike, expiration, isPutOption
* bTokenAmount and collateralMaximum to use in the purchase

The function will then
* Verify the AMM was deployed by our AMM Factory
* Verify the Series doesn't yet exist
* Verify the expiration date and strike are valid for the collateral token
* Create the new series
* Buy the requested options

NOTE
* This will change the operations for the team when deploying new series.  Now each asset will require strike params set and each expiration must be set as allowed before anyone (including the owner) can create a new series.
* The AmmFactory and Erc155Controller address must be set in the AddressesProvider before the auto creation will work.

### CODE REVIEW

File: contracts/series/SeriesController.sol
Comment (MysticDakra) [inline contracts/series/SeriesController.sol:1324]: Do we need to do any check here to make sure we got the correct amount that the user was asking for or is this handled by btokenbuy?

File: contracts/amm/MinterAmm.sol
Comment (MysticDakra) [inline contracts/amm/MinterAmm.sol:1075]: Do we need functions for these getters or can we just query underlyingToken on the amm?  something like amm. underlyingToken since its public?

**Review by MysticDakra — COMMENTED** [review]
Left some general comments I will go over again once we get some test in

File: contracts/series/SeriesController.sol
Comment (siren-tamer) [reply contracts/series/SeriesController.sol]: The aligned date is now checked in the `updateAllowedExpirations` function call.  Then the old create internal function checks to see if the allowed expiration is valid.

File: contracts/series/SeriesController.sol
Comment (siren-tamer) [reply contracts/series/SeriesController.sol:1226]: Yeah, I think Andre has an idea to improve this in the future based off volatility params and automate it instead of setting it explicitly.  For now we said we would just have the "owner" manually set the params for each asset.

File: contracts/series/SeriesController.sol
Comment (siren-tamer) [reply contracts/series/SeriesController.sol]: Check the `createSeriesInternal` function... it checks:
```
require(_expirationDate > block.timestamp, "!expirationDate");
```

File: contracts/series/SeriesController.sol
Comment (siren-tamer) [reply contracts/series/SeriesController.sol]: Yeah the bTokenBuy should error if the amount costs more than `_collateralMaximum` specified.

File: contracts/amm/MinterAmm.sol
Comment (siren-tamer) [reply contracts/amm/MinterAmm.sol:1075]: Yeah this was ugly... the inheritance tree screwed this up since the interface that others reference don't know anything about the public variables in the storage contract.  Keep this comment open and we can investigate moving around the inheritance tree without affecting memory layout is possible.

Comment (seafi) [reply contracts/amm/MinterAmm.sol:1075]: Not sure I fully understand. I'm already accessing this directly for example https://github.com/sirenmarkets/core/blob/4c33b0f164e72a2b1c0f134fafedda37804bf5a6/contracts/amm/AmmDataProvider.sol#L559

File: contracts/series/SeriesController.sol
Comment (seafi) [reply contracts/series/SeriesController.sol:1226]: I think we want the range to be percentage of the underlying price. Otherwise we'd have to update it every time the price moves.

**Comment by seafi** [comment]
@siren-tamer I removed the token getters, seems to be working fine, could you check please

File: contracts/series/SeriesController.sol
Comment (seafi) [inline contracts/series/SeriesController.sol:63]: Should we include the owner role here as well? That way a series can be created manually by the owner multisig as well.

File: contracts/series/SeriesController.sol
Comment (seafi) [reply contracts/series/SeriesController.sol:1226]: Also, I think this logic should live in `SeriesDeployer`. SeriesController doesn't set any limits on what series can be created. It is the SeriesDeployer that is responsible for maintaining the rules.

**Review by MysticDakra — APPROVED** [review]

File: contracts/series/SeriesController.sol
Comment (siren-tamer) [reply contracts/series/SeriesController.sol]: Yep, that change looks good.

**Comment by siren-tamer** [comment]
> @siren-tamer I removed the token getters, seems to be working fine, could you check please

Looks good to me...  for some reason I was getting conflicts previously when I tried that.  Might have messed something up though unrelated.

**Comment by MysticDakra** [comment]
I approved this and I merged in the changes so we dont have merge conflicts now. However there is one test failing can we fix this before we merge this in? I can handle if you want. 

 
      AMM should calculate bToken price correctly
      + expected - actual

      -65273894286428571
      +65274048842857142
      
      at assertBNEq (test/util.ts:228:10)
      at Context.<anonymous> (test/amm/minterAmmCall.ts:1048:15)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

**Comment by seafi** [comment]
> 1048

let's fix any flaky tests separately. This is not related to this issue, just need to add precise timing there.

**Pull request merged**
